File: python/test/clamp.py
import torch
import torch.nn as nn
import torch.nn.functional as F

import torch_tpu
device = "tpu"
def case1():
    a = torch.randn((5,5))
    a_tpu = a.to(device)
    
    o_cpu = torch.clamp(a, max=1, min=-1)
    o_tpu = torch.clamp(a_tpu, max=1, min=-1)
    print("cpu: ", o_cpu)
    print("tpu: ", o_tpu.cpu())

if __name__ == "__main__":
    case1()

File: libtorch_plugin/ops/Clamp.cpp
#include <torch/library.h>
#include <torch/torch.h>
#include <ATen/core/TensorBase.h>
#include <ATen/EmptyTensor.h>
#include <ATen/native/ConvUtils.h>
#include <TPUDeviceManager.h>
#include <TPUTorchUtils.h>
#include <sgdnn_api.h>
#include "common/config.h"

namespace at {
Tensor & clamp_out_tpu( const at::Tensor & self, const c10::optional<at::Scalar> & min,
                        const c10::optional<at::Scalar> & max, at::Tensor & out) {
  CHECK_TENSOR_IN_DEVICE ( self );
  CHECK_TENSOR_IN_DEVICE ( out );
#if 1
    auto out_cpu = clamp ( self.to(torch::kFloat32).cpu(), min, max );
    out = out_cpu.to(out.device()).to(out.dtype());
#endif

    return out;
}

TORCH_LIBRARY_IMPL ( aten, TPU, m )
{
 m.impl ( "clamp.out",  clamp_out_tpu);
}

} // namespace at